File: src/pomodoro/widget/PomodoroWidget.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package pomodoro.widget;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.wm.CustomStatusBarWidget;
import com.intellij.openapi.wm.StatusBar;
import com.intellij.openapi.wm.StatusBarWidget;
import com.intellij.util.ui.UIUtil;
import org.jetbrains.annotations.NotNull;
import pomodoro.PomodoroComponent;
import pomodoro.UIBundle;
import pomodoro.model.ChangeListener;
import pomodoro.model.PomodoroModel;
import pomodoro.model.Settings;
import pomodoro.toolkitwindow.PomodoroPresenter;

import javax.swing.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

import static pomodoro.PomodoroComponent.getSettings;

public class PomodoroWidget implements CustomStatusBarWidget, StatusBarWidget.Multiframe, ChangeListener {
	private final ImageIcon pomodoroIcon = new ImageIcon(getClass().getResource("/resources/pomodoro.png"));
	private final ImageIcon pomodoroStoppedIcon = new ImageIcon(getClass().getResource("/resources/pomodoroStopped.png"));
	private final ImageIcon pomodoroBreakIcon = new ImageIcon(getClass().getResource("/resources/pomodoroBreak.png"));
	private final ImageIcon pomodoroDarculaIcon = new ImageIcon(getClass().getResource("/resources/pomodoro-inverted.png"));
	private final ImageIcon pomodoroStoppedDarculaIcon = new ImageIcon(getClass().getResource("/resources/pomodoroStopped-inverted.png"));
	private final ImageIcon pomodoroBreakDarculaIcon = new ImageIcon(getClass().getResource("/resources/pomodoroBreak-inverted.png"));
	private final TextPanelWithIcon panelWithIcon;
	private StatusBar statusBar;
	private final PomodoroModel model;


	public PomodoroWidget() {
		panelWithIcon = new TextPanelWithIcon();
		final Settings settings = getSettings();

		PomodoroComponent pomodoroComponent = ApplicationManager.getApplication().getComponent(PomodoroComponent.class);
		assert pomodoroComponent != null;
		model = pomodoroComponent.getModel();
		updateWidgetPanel(model, panelWithIcon, settings.isShowTimeInToolbarWidget());

		model.addUpdateListener(panelWithIcon, new Runnable() {
			@Override
			public void run() {
				SwingUtilities.invokeLater(new Runnable() {
					@Override
					public void run() {
						updateWidgetPanel(model, panelWithIcon, settings.isShowTimeInToolbarWidget());
					}
				});
			}
		});
		panelWithIcon.addMouseListener(new MouseAdapter() {
			@Override
			public void mouseClicked(MouseEvent e) {
				model.onUserSwitchToNextState();
			}

			@Override
			public void mouseEntered(MouseEvent e) {
				String tooltipText = tooltipText(model);
				statusBar.setInfo(tooltipText);
			}

			@Override
			public void mouseExited(MouseEvent e) {
				statusBar.setInfo("");
			}
		});
	}

	@Override
	public void install(@NotNull StatusBar statusBar) {
		this.statusBar = statusBar;
	}

	@Override
	public JComponent getComponent() {
		return panelWithIcon;
	}

	@Override
	public StatusBarWidget copy() {
		return new PomodoroWidget();
	}

	private String tooltipText(PomodoroModel model) {
		String nextAction = nextActionName(model);
		int pomodorosAmount = model.getPomodorosAmount();
		return UIBundle.message("statuspanel.tooltip", nextAction, pomodorosAmount);
	}

	@NotNull
	private static String nextActionName(PomodoroModel model) {
		switch (model.getState()) {
			case STOP: return UIBundle.message("statuspanel.start");
			case RUN: return UIBundle.message("statuspanel.stop");
			case BREAK: return UIBundle.message("statuspanel.stop_break");
			default: return "";
		}
	}

	private void updateWidgetPanel(PomodoroModel model, TextPanelWithIcon panelWithIcon, boolean showTimeInToolbarWidget) {
		if (showTimeInToolbarWidget) {
			int timeLeft = model.getProgressMax() - model.getProgress();
			panelWithIcon.setText(PomodoroPresenter.formatTime(timeLeft));
		} else {
			panelWithIcon.setText("");
		}
		panelWithIcon.setIcon(getIcon(model));
		panelWithIcon.repaint();
	}

	@NotNull
	private ImageIcon getIcon(PomodoroModel model) {
		boolean underDarcula = UIUtil.isUnderDarcula();
		switch (model.getState()) {
			case STOP: return underDarcula ? pomodoroStoppedDarculaIcon : pomodoroStoppedIcon;
			case RUN: return underDarcula ? pomodoroDarculaIcon : pomodoroIcon;
			case BREAK: return underDarcula ? pomodoroBreakDarculaIcon : pomodoroBreakIcon;
			default: throw new IllegalStateException();
		}
	}

	@Override
	public WidgetPresentation getPresentation(@NotNull PlatformType type) {
		return null;
	}

	@Override
	public void dispose() {
	}

	@NotNull
	@Override
	public String ID() {
		return "Pomodoro";
	}

	@Override
	public void onChange(Settings settings) {
		updateWidgetPanel(model, panelWithIcon, settings.isShowTimeInToolbarWidget());
	}
}

File: src/pomodoro/widget/PomodoroWidget.kt
package pomodoro.widget

import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.actionSystem.impl.AsyncDataContext
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.ui.popup.JBPopupFactory.ActionSelectionAid.SPEEDSEARCH
import com.intellij.openapi.wm.CustomStatusBarWidget
import com.intellij.openapi.wm.StatusBar
import com.intellij.openapi.wm.StatusBarWidget
import com.intellij.openapi.wm.StatusBarWidgetFactory
import com.intellij.ui.JBColor
import com.intellij.ui.awt.RelativePoint
import pomodoro.PomodoroService
import pomodoro.ResetPomodorosCounter
import pomodoro.StartOrStopPomodoro
import pomodoro.UIBundle
import pomodoro.model.PomodoroModel
import pomodoro.model.PomodoroSnapshot
import pomodoro.model.PomodoroState
import pomodoro.model.PomodoroState.Mode.*
import pomodoro.model.Settings
import pomodoro.model.time.Duration
import pomodoro.model.time.Time
import java.awt.Point
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.ImageIcon
import javax.swing.JComponent

class PomodoroWidgetFactory : StatusBarWidgetFactory {
    override fun getId() = "Pomodoro"
    override fun getDisplayName() = "Pomodoro"
    override fun isAvailable(project: Project) = true
    override fun createWidget(project: Project) = PomodoroWidget()
    override fun disposeWidget(widget: StatusBarWidget) = widget.dispose()
    override fun canBeEnabledOn(statusBar: StatusBar) = true
}

class PomodoroWidget : CustomStatusBarWidget, StatusBarWidget.Multiframe, Settings.ChangeListener {
    private val panelWithIcon = TextPanelWithIcon()
    private lateinit var statusBar: StatusBar
    private val model = service<PomodoroService>().model

    init {
        val settings = service<Settings>()
        updateWidgetPanel(model, panelWithIcon, settings.isShowTimeInToolbarWidget)

        model.addListener(this, object : PomodoroModel.Listener {
            override fun onStateChange(state: PomodoroState, wasManuallyStopped: Boolean) {
                ApplicationManager.getApplication().invokeLater { updateWidgetPanel(model, panelWithIcon, settings.isShowTimeInToolbarWidget) }
            }
        })
        panelWithIcon.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(event: MouseEvent?) {
                if (event == null) return
                if (event.isAltDown) {
                    val popup = JBPopupFactory.getInstance().createActionGroupPopup(
                        null,
                        DefaultActionGroup(listOf(StartOrStopPomodoro(), ResetPomodorosCounter())),
                        MapDataContext(mapOf(PlatformDataKeys.CONTEXT_COMPONENT.name to event.component)),
                        SPEEDSEARCH,
                        true
                    )
                    val dimension = popup.content.preferredSize
                    val point = Point(0, -dimension.height)
                    popup.show(RelativePoint(event.component, point))
                } else if (event.button == MouseEvent.BUTTON1) {
                    model.onUserSwitchToNextState(Time.now())
                }
            }

            override fun mouseEntered(e: MouseEvent?) {
                val (today, week, month) = PomodoroSnapshot.statisticsAt(Time.now(), model.state.history)
                statusBar.info = UIBundle.message("statuspanel.tooltip", nextActionName(model),
                        today.completed, today.failed, week.completed, week.failed, month.completed, month.failed)
            }

            override fun mouseExited(e: MouseEvent?) {
                statusBar.info = ""
            }
        })
    }

    override fun install(statusBar: StatusBar) {
        this.statusBar = statusBar
    }

    override fun getComponent(): JComponent = panelWithIcon

    override fun copy() = PomodoroWidget()

    private fun nextActionName(model: PomodoroModel) = when (model.state.mode) {
        Run   -> UIBundle.message("statuspanel.stop")
        Break -> UIBundle.message("statuspanel.stop_break")
        Stop  -> UIBundle.message("statuspanel.start")
    }

    private fun updateWidgetPanel(model: PomodoroModel, panelWithIcon: TextPanelWithIcon, showTimeInToolbarWidget: Boolean) {
        panelWithIcon.text = if (showTimeInToolbarWidget) model.timeLeft.formatted() else ""
        panelWithIcon.toolTipText = UIBundle.message("statuspanel.widget_tooltip", nextActionName(model))
        panelWithIcon.icon = when (model.state.mode) {
            Run   -> if (JBColor.isBright()) pomodoroIcon else pomodoroDarculaIcon
            Break -> if (JBColor.isBright()) pomodoroBreakIcon else pomodoroBreakDarculaIcon
            Stop  -> if (JBColor.isBright()) pomodoroStoppedIcon else pomodoroStoppedDarculaIcon
        }
        panelWithIcon.repaint()
    }

    override fun dispose() {
        model.removeListener(this)
    }

    override fun ID() = "Pomodoro"

    override fun onChange(newSettings: Settings) {
        updateWidgetPanel(model, panelWithIcon, newSettings.isShowTimeInToolbarWidget)
    }

    private fun Duration.formatted(): String {
        val formattedMinutes = String.format("%02d", minutes)
        val formattedSeconds = String.format("%02d", (this - Duration(minutes)).seconds)
        return "$formattedMinutes:$formattedSeconds"
    }

    private class MapDataContext(private val map: Map<String, Any?> = HashMap()) : AsyncDataContext {
        override fun getData(dataId: String) = map[dataId]
    }

    companion object {
        private val pomodoroIcon = loadIcon("/pomodoro.png")
        private val pomodoroStoppedIcon = loadIcon("/pomodoroStopped.png")
        private val pomodoroBreakIcon = loadIcon("/pomodoroBreak.png")
        private val pomodoroDarculaIcon = loadIcon("/pomodoro_dark.png")
        private val pomodoroStoppedDarculaIcon = loadIcon("/pomodoroStopped_dark.png")
        private val pomodoroBreakDarculaIcon = loadIcon("/pomodoroBreak_dark.png")

        private fun loadIcon(filePath: String) = ImageIcon(PomodoroWidget::class.java.getResource(filePath))
    }
}
